--[=[
	@class LuauBuiltinFunction

	Builtin function ids, used in LOP_FASTCALL
]=]
return {
	LBF_NONE = 0,

	-- assert()
	LBF_ASSERT = 1,

	-- math.
	LBF_MATH_ABS = 2,
	LBF_MATH_ACOS = 3,
	LBF_MATH_ASIN = 4,
	LBF_MATH_ATAN2 = 5,
	LBF_MATH_ATAN = 6,
	LBF_MATH_CEIL = 7,
	LBF_MATH_COSH = 8,
	LBF_MATH_COS = 9,
	LBF_MATH_DEG = 10,
	LBF_MATH_EXP = 11,
	LBF_MATH_FLOOR = 12,
	LBF_MATH_FMOD = 13,
	LBF_MATH_FREXP = 14,
	LBF_MATH_LDEXP = 15,
	LBF_MATH_LOG10 = 16,
	LBF_MATH_LOG = 17,
	LBF_MATH_MAX = 18,
	LBF_MATH_MIN = 19,
	LBF_MATH_MODF = 20,
	LBF_MATH_POW = 21,
	LBF_MATH_RAD = 22,
	LBF_MATH_SINH = 23,
	LBF_MATH_SIN = 24,
	LBF_MATH_SQRT = 25,
	LBF_MATH_TANH = 26,
	LBF_MATH_TAN = 27,

	-- bit32.
	LBF_BIT32_ARSHIFT = 28,
	LBF_BIT32_BAND = 29,
	LBF_BIT32_BNOT = 30,
	LBF_BIT32_BOR = 31,
	LBF_BIT32_BXOR = 32,
	LBF_BIT32_BTEST = 33,
	LBF_BIT32_EXTRACT = 34,
	LBF_BIT32_LROTATE = 35,
	LBF_BIT32_LSHIFT = 36,
	LBF_BIT32_REPLACE = 37,
	LBF_BIT32_RROTATE = 38,
	LBF_BIT32_RSHIFT = 39,

	-- type()
	LBF_TYPE = 40,

	-- string.
	LBF_STRING_BYTE = 41,
	LBF_STRING_CHAR = 42,
	LBF_STRING_LEN = 43,

	-- typeof()
	LBF_TYPEOF = 44,

	-- string.
	LBF_STRING_SUB = 45,

	-- math.
	LBF_MATH_CLAMP = 46,
	LBF_MATH_SIGN = 47,
	LBF_MATH_ROUND = 48,

	-- raw*
	LBF_RAWSET = 49,
	LBF_RAWGET = 50,
	LBF_RAWEQUAL = 51,

	-- table.
	LBF_TABLE_INSERT = 52,
	LBF_TABLE_UNPACK = 53,

	-- vector ctor
	LBF_VECTOR = 54,

	-- bit32.count
	LBF_BIT32_COUNTLZ = 55,
	LBF_BIT32_COUNTRZ = 56,

	-- select(_, ...)
	LBF_SELECT_VARARG = 57,

	-- rawlen
	LBF_RAWLEN = 58,

	-- bit32.extract(_, k, k)
	LBF_BIT32_EXTRACTK = 59,

	-- get/setmetatable
	LBF_GETMETATABLE = 60,
	LBF_SETMETATABLE = 61,

	-- tonumber/tostring
	LBF_TONUMBER = 62,
	LBF_TOSTRING = 63,

	-- bit32.byteswap(n)
	LBF_BIT32_BYTESWAP = 64,

	-- buffer.
	LBF_BUFFER_READI8 = 65,
	LBF_BUFFER_READU8 = 66,
	LBF_BUFFER_WRITEU8 = 67,
	LBF_BUFFER_READI16 = 68,
	LBF_BUFFER_READU16 = 69,
	LBF_BUFFER_WRITEU16 = 70,
	LBF_BUFFER_READI32 = 71,
	LBF_BUFFER_READU32 = 72,
	LBF_BUFFER_WRITEU32 = 73,
	LBF_BUFFER_READF32 = 74,
	LBF_BUFFER_WRITEF32 = 75,
	LBF_BUFFER_READF64 = 76,
	LBF_BUFFER_WRITEF64 = 77,
}